Implement parental controls for profiles

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/TNL-1606

@cahrens @dan-f Please review.

andy-armstrong commented:
Note for reviewers: we decided not to physically remove profile images as part of this task. Instead we just remove the indication that they have an image, and we will decide at a later point how (or if) to clean up these unreferenced images.

Ferdi commented:
Can you please link to some description of what this is suppose to do ?
thanks !

andy-armstrong commented:
Sorry @Ferdi. I've replaced our JIRA ticket number with a link. We are implementing a simple parental control system where users under thirteen will not be able to share their profiles or upload profile images. The longer term goal is to add true parental controls so that a parent can allow their child to have a public profile.

cahrens commented on common/djangoapps/student/models.py:
```diff
@@ -284,6 +279,46 @@ def set_login_session(self, session_id=None):
+            date (Date): The date for which consent needs to be tested (defaults to now).
+            age_limit (int): The age limit at which parental consent is no longer required.
+                This defaults to the value of the setting 'PARENTAL_CONTROL_AGE_LIMIT'.
+            default_requires_consent (Bool): True if users require parental consent if they
```
Nit: I believe it is bool instead of Bool

cahrens commented on common/djangoapps/student/models.py:
```diff
@@ -284,6 +279,46 @@ def set_login_session(self, session_id=None):
+            return default_requires_consent
+        if date is None:
+            date = datetime.now(UTC)
+        return date.year - year_of_birth <=  age_limit # pylint: disable=maybe-no-member
```
<= limit to be conservative enough in case the birthdate month is later in the year than the current month?

I think there is an extra space after <=, btw.

andy-armstrong replied:
Yes, exactly. I'll add a comment to make that explicit, and I'll fix the extra space.

cahrens commented on common/djangoapps/student/tests/test_parental_controls.py:
```diff
@@ -0,0 +1,87 @@
+
+        # verify that a user's profile image is removed when they switch to requiring parental controls
+        self.set_year_of_birth(current_year - 10)
+        self.profile.has_profile_image = True
```
Nit: the test seems clearer to me if you delete line 85 (since the user had a profile image before you changed the year_of_birth).

cahrens commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -235,33 +230,28 @@ def update_email_opt_in(user, org, optin):
+    Raises:
+         UserNotFound: no user profile exists for the specified user.
     """
-    # Avoid calling get_account_settings because it introduces circularity for many callers who need both
```
You removed the comment, but didn't remove the user_profile object on line 237. Is user_profile still needed?

cahrens commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -235,33 +230,28 @@ def update_email_opt_in(user, org, optin):
     )
-
+    user_requires_parental_consent = user_profile.requires_parental_consent(
+        age_limit=getattr(settings, 'EMAIL_OPTIN_MINIMUM_AGE', 13),
```
I'm trying to work out if this default value of 13 should be here. Is this just for backwards compatibility?

andy-armstrong replied:
I left the 13 because it was there in the old code, and if it defaults to None then parental controls get switched off. It does feel a little odd to have the 13 left sprinkled through the code, so I'd be happy to remove it if that seems reasonable.

cahrens replied:
I don't feel strongly about it... Do whichever you feel most comfortable with.

andy-armstrong replied:
I put the calculation of user_requires_parental_consent until immediately before it was needed. Your point is a good one that it is the only user of the user_profile, so that should all be grouped together. I'll shuffle the code around to make it more logical.

cahrens commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -235,33 +230,28 @@ def update_email_opt_in(user, org, optin):
-
     except IntegrityError as err:
         log.warn(u"Could not update organization wide preference due to IntegrityError: {}".format(err.message))
+    if settings.FEATURES.get('SEGMENT_IO_LMS') and settings.SEGMENT_IO_LMS_KEY:
```
Why did you move this out of the try? I would imagine we shouldn't track the event if the save failed.

andy-armstrong replied:
Good point. I was trying to scope the exception handling more tightly, because it was previously around all sorts of code that shouldn't be raising that exception. I'll move the event handling back in.

andy-armstrong commented:
@cahrens Thanks for the review. I've pushed up changes to address your feedback.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_helpers.py:
```diff
@@ -26,6 +28,7 @@ class ProfileImageUrlTestCase(TestCase):
     def setUp(self):
         super(ProfileImageUrlTestCase, self).setUp()
         self.user = UserFactory()
+        self.user.profile.year_of_birth = 1980    # Ensure that parental controls don't apply to this user
```
I think you should save the user/profile thingy after updating it.

cahrens commented:
:+1:

cahrens commented on common/djangoapps/student/models.py:
```diff
@@ -284,6 +279,50 @@ def set_login_session(self, session_id=None):
+
+        Args:
+            date (Date): The date for which consent needs to be tested (defaults to now).
+            age_limit (int): The age limit at which parental consent is no longer required.
```
@andy-armstrong Oddly enough, you missed this usage of "Bool" when changing to "bool".

andy-armstrong replied:
Whoops! I only went through user_api...

andy-armstrong commented:
@cahrens I've updated all the incorrect docstrings in the student Django app including the one I newly introduced. I've squashed and repushed since I figure you don't need to re-review.

dan-f commented:
:+1: lgtm!
